---
title: "UTR Length and Alternative Trans-splicing / Poly-adenylation Analysis"
output:
  knitrBootstrap::bootstrap_document:
    theme: flatly
---

```{r knitr_settings, include=FALSE}
library(knitr)
opts_chunk$set(fig.width=1920/96,
               fig.height=1080/96,
               dpi=96)
options(knitr.duplicate.label='allow')
```

Overview
--------

The goal of this analysis is to parse the output from our [UTR analysis
pipeline](https://github.com/elsayed-lab/utr_analysis), perform some basic
smoothing of the spliced leader and polya acceptor site counts and determine
the most likely primary UTR boundaries for each gene where information is
available. 5'- and 3'-UTR boundaries, along with some other useful information
such as UTR GC- and CT-richness will be outputted in a format that is
convenient for downstream analysis.

Finally, we will look for evidence of either alternative trans-splicing or
poly-adenylation and attempt to visualize the prevelance and magnitude of
these events across the different developmental stages.

This analysis is based on a version originally written and applied separately
to *T. cruzi* and *L. major* in 2014. The code has been refactored and cleaned
up to make it easier to apply to new datasets.

```{r run_date, results='asis', echo=FALSE}
email = "<a href='mailto:khughitt@umd.edu'>Keith Hughitt</a>"
last_update = format(Sys.time(), "(<time>%Y-%m-%d</time>)")
cat(paste(email, last_update))

# Clean up any existing variables
rm(list=ls())

# Create output and build directories if needed
for (x in c('build', 'output')) {
    if (!file.exists(x)) {
        dir.create(x)
    }
}
```

Settings
--------

```{r settings}
# Number of individual UTR plots to display for each developmental stage and
# UTR side. The purpose of these plots are to provide a sense of the data
# distribution and to visualize the effect of primary # site selection with and
# without smoothing.
max_plots = 10
```

[view source](README.rmd)

Methods
-------

### Libraries

```{r load_libraries, warnings=FALSE}
library(GenomicRanges)
library(Biostrings)
library(rtracklayer)
library(ggplot2)
library(ggvis)
library(reshape2)
library(dplyr)
source('helper.R')
```

### Settings

```{r child='settings/tcruzi_sylvio.Rmd'}
```

### Load sequence and annotations

```{r load_annotations, warning=FALSE}
# Load genome sequence and annotations
gff = import.gff3(input_gff)
fasta = readDNAStringSet(input_fasta)

chromosomes = gff[gff$type %in% c('contig', 'chromosome')]
genes       = gff[gff$type == 'gene']

# L. major: load unannotated ORFs detected from ribosome profiling data
if (parasite == 'L. major') {
    orfs      = gff[gff$type == 'ORF']
    orfs$Name = orfs$ID
    orfs$description = 'Unannotated ORF'

    # 2015/06/30 A few of the unannotated ORFs appear to have multiple conflicting
    # entries -- removing these for now...
    orfs = orfs[!duplicated(orfs$Name)]

    # Drop GFF columns not shared betwene TriTrypDB GFF and uORF GFF
    keep_cols = intersect(colnames(mcols(genes)), colnames(mcols(orfs)))
    genes = genes[,keep_cols]
    orfs = orfs[,keep_cols]

    genes = append(genes, orfs)

    # Fix names (L. major chromosome identifiers)
    names(fasta) = substring(names(fasta), 0, 7)
}
```

### Remove noncoding RNAs

```{r filter_ncrnas}
# filter noncoding RNAs
noncoding_ids = genes$ID[grepl(id_filter_string, genes$ID)]

ncrna_mask = rep(FALSE, length(genes))

if (!is.null(id_filter_string)) {
    ncrna_mask = ncrna_mask | grepl(id_filter_string, genes$ID)
}

if (!is.null(type_filter_string)) {
    ncrna_mask = ncrna_mask | grepl(type_filter_string, as.character(genes$type))
}

# remove noncoding RNAs
genes = genes[!ncrna_mask,]

# remaining gene ids
gene_ids = genes$Name
```

### Load spliced leader and poly(A) site coordinates

```{r load_acceptor_sites}
sl_gffs = list()
polya_gffs = list()

# names of developmental stages being analyzed
stages = names(sl_input)

for (stage in stages) {
    # SL acceptor sites
    sl_gffs[stage] = import.gff3(gzfile(sl_input[[stage]]))
    sl_gffs[stage] = sl_gffs[[stage]][sl_gffs[[stage]]$Name %in% gene_ids]

    # Poly(A) sites
    polya_gffs[stage] = import.gff3(gzfile(polya_input[[stage]]))
    polya_gffs[stage] = polya_gffs[[stage]][polya_gffs[[stage]]$Name %in% gene_ids]
}

# Normalize seqinfo (e.g. when reads cover differing sets of contigs)
sl_seqinfo = seqinfo(sl_gffs[[stages[1]]])
polya_seqinfo = seqinfo(polya_gffs[[stages[1]]])

for (stage in stages[2:length(stages)]) {
    sl_seqinfo = suppressWarnings(merge(sl_seqinfo, seqinfo(sl_gffs[[stage]])))
    polya_seqinfo = suppressWarnings(merge(polya_seqinfo, seqinfo(polya_gffs[[stage]])))
}

for (stage in stages) {
    # Add levels that aren't already represented
    sl_levels = seqlevels(sl_seqinfo)[!seqlevels(sl_seqinfo) 
                                        %in% seqlevels(sl_gffs[[stage]])]
    seqlevels(sl_gffs[[stage]]) = c(seqlevels(sl_gffs[[stage]]), sl_levels)

    # Add levels that aren't already represented
    polya_levels = seqlevels(polya_seqinfo)[!seqlevels(polya_seqinfo) 
                                        %in% seqlevels(polya_gffs[[stage]])]
    seqlevels(polya_gffs[[stage]]) = c(seqlevels(polya_gffs[[stage]]), 
                                       polya_levels)
}
```

### Methods and Results

#### 5'UTR

```{r utr5_settings, include=FALSE}
feature_name = 'sl'
gffs         = sl_gffs
default_utr_width = default_5utr_width
```

```{r child='child/utr_length.Rmd'}
```

```{r child='child/utr_composition.Rmd'}
```

```{r child='child/results.Rmd'}
```

#### 3'UTR

```{r utr3_settings, include=FALSE}
gffs         = polya_gffs
feature_name = 'polya'
default_utr_width = default_3utr_width
```

```{r child='child/utr_length.Rmd'}
```

```{r child='child/utr_composition.Rmd'}
```

```{r child='child/results.Rmd'}
```

System Information
------------------

```{r sysinfo}
sessionInfo()
date()
```
