```{r settings}
#
# Similar to lmajor_all.Rmd except that instead of reading in individual
# species and stage-specific GFFs, combined GFFs representing sites detecting
# across all samples are used.
#
# In theory the results should be similar, although the improved read depth
# may improve the site detection some in the previous utr analysis pipeline.
#
# species information
parasite = 'L. major'
host     = 'H. sapiens - M. Musculus'
stages  = c('lmajor')

# output file prefix
outfile_prefix = 'lmajor_all_combined'

# fasta/gff
input_gff   = file.path(Sys.getenv('REF'), 'lmajor_friedlin',
                        'annotation/TriTrypDB-8.1_LmajorFriedlin_plus_unannotated_orfs.gff')
input_fasta = file.path(Sys.getenv("REF"), 
                        'lmajor_friedlin/genome/TriTrypDB-8.1_LmajorFriedlin_Genome.fasta')

# spliced leader / polyadeynlation sites
# Taken from L. major / H. sapiens  UTR analysis output from 2015/05/20
sl_input = list(
    lmajor='input/lmajor_all_no_ncrnas_sl.gff.gz'
)

polya_input = list(
    lmajor='input/lmajor_all_no_ncrnas_polya.gff.gz'
)

# Number of bases flanking CDS to scan for motifs when actual UTR length is not
# known; numbers are based on median 5' and 3' UTR lengths for L. major
default_5utr_width = 250
default_3utr_width = 545

# Non-coding RNA filters
id_filter_string   = 'rRNA|5SRRNA|snRNA|snoRNA|SLRNA|TRNA|SRP'
type_filter_string = 'rRNAencoding|snRNAencoding|snoRNAencoding|tRNAencoding'
```
