```{r settings}
# species information
parasite = 'T. cruzi strain Sylvio'
host     = 'H. sapiens'
stages  = c('trypomastigote', 'amastigote')

# output file prefix
outfile_prefix = 'tcruzi_sylvio'

# fasta/gff
input_gff   = file.path(Sys.getenv('REF'), 'tcruzi_sylvio',
                        '/annotation/TriTrypDB-8.0_TcruziSylvioX10-1.gff')
input_fasta = file.path(Sys.getenv("REF"), 
                        'tcruzi_sylvio/genome/TriTrypDB-8.0_TcruziSylvioX10-1_Genome.fasta')

# spliced leader / polyadeynlation sites
# Taken from T. cruzi Sylvio UTR analysis output from 2015/07
sl_input = list(
    amastigote='input/tcruzi_sylvio_amastigote_spliced_leader_sorted.gff.gz',
    trypomastigote='input/tcruzi_sylvio_trypomastigote_spliced_leader_sorted.gff.gz'
)

polya_input = list(
    amastigote='input/tcruzi_sylvio_amastigote_polya_sorted.gff.gz',
    trypomastigote='input/tcruzi_sylvio_trypomastigote_polya_sorted.gff.gz'
)

# Number of bases flanking CDS to scan for motifs when actual UTR length is not
# known; numbers are based on median 5' and 3' UTR lengths for T. cruzi Sylvio
default_5utr_width = 250
default_3utr_width = 575

# Non-coding RNA filters
id_filter_string   = 'rRNA|5SRRNA|snRNA|snoRNA|SLRNA|TRNA|SRP'
type_filter_string = 'rRNAencoding|snRNAencoding|snoRNAencoding|tRNAencoding'
```
