```{r settings}
# species information
parasite = 'T. cruzi strain Sylvio / Y. strain'
host     = 'H. sapiens'
stages  = c('ystrain', 'sylvio')

# Output verbosity
verbose = TRUE

# Number of individual UTR plots to display for each developmental stage and
# UTR side. The purpose of these plots are to provide a sense of the data
# distribution and to visualize the effect of primary # site selection with and
# without smoothing.
max_plots = 10

# output file prefix
outfile_prefix = 'tcruzi_sylvio_vs_ystrain_72hrs'

# fasta/gffs
sylvio_gff = file.path(Sys.getenv('REF'), 'tcruzi_sylvio',
                        '/annotation/TriTrypDB-8.0_TcruziSylvioX10-1.gff')
sylvio_fasta = file.path(Sys.getenv("REF"), 
                        'tcruzi_sylvio/genome/TriTrypDB-8.0_TcruziSylvioX10-1_Genome.fasta')

ystrain_gff   = file.path(Sys.getenv('REF'), 'tcruzi_clbrener_esmeraldo-like',
                        '/annotation/TriTrypDB-8.1_TcruziCLBrenerEsmeraldo-like.gff')
ystrain_fasta = file.path(Sys.getenv("REF"), 
                        'tcruzi_clbrener_esmeraldo-like/genome/TriTrypDB-8.1_TcruziCLBrenerEsmeraldo-like_Genome.fasta')

input_gffs = list(
    sylvio=sylvio_gff,
    ystrain=ystrain_gff
)

input_fastas = list(
    sylvio=sylvio_fasta,
    ystrain=ystrain_fasta
)

# spliced leader / polyadeynlation sites
# Taken from T. cruzi Sylvio UTR analysis output from 2016/02
sl_input = list(
    sylvio='input/tcruzi_sylvio_72hrs_sl_sorted.gff.gz',
    ystrain='input/tcruzi_ystrain_72hrs_sl_sorted.gff.gz'
)

polya_input = list(
    sylvio='input/tcruzi_sylvio_72hrs_polya_sorted.gff.gz',
    ystrain='input/tcruzi_ystrain_72hrs_polya_sorted.gff.gz'
)

# Number of bases flanking CDS to scan for motifs when actual UTR length is not
# known; numbers are based on median 5' and 3' UTR lengths for T. cruzi Sylvio
default_5utr_width = 95
default_3utr_width = 270

# Non-coding RNA filters
id_filter_string   = 'rRNA|5SRRNA|snRNA|snoRNA|SLRNA|TRNA|SRP'
type_filter_string = 'rRNAencoding|snRNAencoding|snoRNAencoding|tRNAencoding'
```
